File: Audiograph/ArtistAdvanced.vb
Public Class frmArtistAdvanced
    Private Sub Search(sender As Object, e As EventArgs) Handles btnSearch.Click
        ' end if no data has been entered in track box
        If txtSearchArtist.Text = String.Empty Then
            MessageBox.Show("Please make sure you have entered data in the artist search field.", "Advanced Search", MessageBoxButtons.OK, MessageBoxIcon.Warning)
            Exit Sub
        End If

        ' get search results
        Dim results As String = CallAPI("artist.search", "", "artist=" & txtSearchArtist.Text.Trim.Replace(" ", "+"))

        ' remove opensearch garbage
        If results.Contains("opensearch") = True Then
            Dim startindex As UInteger = InStr(results, "<opensearch:Query") - 1            ' find starting index
            Dim endindex As UInteger = InStr(results, "</opensearch:itemsPerPage>") + 25    ' find ending index
            results = results.Remove(startindex, endindex - startindex)                     ' remove (get number of chars between start and end index)
        End If

        ' check for errors/populate listview
        ltvResults.Items.Clear()
        ' get status
        Dim status As String = ParseMetadata(results, "lfm status=")
        If status.Contains("ok") = True Then
            For count As Byte = 0 To 29
                ' parse for data
                Dim searchnodes() As String = {"name", "listeners"}
                ParseXML(results, "/lfm/results/artistmatches/artist", count, searchnodes)

                ' add to listview
                If searchnodes(0).Contains("ERROR: ") = False Then
                    ltvResults.Items.Add(searchnodes(0)) ' name
                    ltvResults.Items(count).SubItems.Add(CUInt(searchnodes(1)).ToString("N0"))  ' listeners
                End If
            Next count
        Else    ' if errors
            MessageBox.Show("API ERROR: Cannot retrieve search results.", "Advanced Search", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Exit Sub
        End If

        ' display message if no results found
        If ltvResults.Items.Count = 0 Then
            MessageBox.Show("No results found, please alter your search and try again.", "Search: " & txtSearchArtist.Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
            txtSearchArtist.Select()
        End If
    End Sub

    ' put selected search item into final boxes
    Private Sub SelectSearchItem(sender As Object, e As ListViewItemSelectionChangedEventArgs) Handles ltvResults.ItemSelectionChanged
        If ltvResults.SelectedItems.Count > 0 Then
            txtArtist.Text = ltvResults.SelectedItems(0).Text
            txtListeners.Text = ltvResults.SelectedItems(0).SubItems(1).Text
        End If
    End Sub

    Private Sub MBID(sender As Object, e As EventArgs) Handles btnMBID.Click
        ' call api
        Dim info As String = CallAPI("artist.getInfo", "", "mbid=" & txtMBID.Text.Trim.ToLower)

        ' get info or error
        Dim infonodes() As String = {"name", "stats/listeners"}
        ParseXML(info, "/lfm/artist", 0, infonodes)
        ' check for error
        If infonodes(0).Contains("ERROR: ") = False AndAlso infonodes(0).Contains("ERROR: ") = False Then
            txtArtist.Text = infonodes(0)
            Try
                txtListeners.Text = CInt(infonodes(1)).ToString("N0")
            Catch ex As Exception
                txtListeners.Text = "Number Error"
            End Try
        Else
            MessageBox.Show("No artist was able to be found from this MBID, please try again.", "MBID: " & txtMBID.Text.Trim.ToLower, MessageBoxButtons.OK, MessageBoxIcon.Error)
            txtMBID.Select()
        End If
    End Sub

    ' put final box data into main form
    Private Sub OK(sender As Object, e As EventArgs) Handles btnOK.Click
        If txtArtist.Text <> String.Empty AndAlso txtArtist.Text <> "N/A" Then   ' if there is data in artist box
            GoToArtist(txtArtist.Text.Trim)
            Me.Close()                                      ' close window
        Else  ' if there is no data entered
            Me.Close()
        End If
    End Sub

    Private Sub Cancel(sender As Object, e As EventArgs) Handles btnCancel.Click
        Me.Close()
    End Sub

    Private Sub UserType(sender As TextBox, e As KeyEventArgs) Handles txtArtist.KeyDown
        If ltvResults.SelectedItems.Count = 0 Then
            Exit Sub
        End If

        ' clear all info
        ltvResults.SelectedItems.Clear()
        txtListeners.Text = "N/A"
        sender.Clear()
    End Sub

    Private Sub DoubleClickItem(sender As Object, e As EventArgs) Handles ltvResults.ItemActivate
        btnOK.PerformClick()
    End Sub
End Class

File: Audiograph/ArtistAdvanced.cs
using System;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace Audiograph;

public partial class frmArtistAdvanced
{
    public frmArtistAdvanced()
    {
        InitializeComponent();
    }

    private void Search(object sender, EventArgs e)
    {
        // end if no data has been entered in track box
        if (string.IsNullOrEmpty(txtSearchArtist.Text))
        {
            MessageBox.Show("Please make sure you have entered data in the artist search field.", "Advanced Search",
                MessageBoxButtons.OK, MessageBoxIcon.Warning);
            return;
        }

        // get search results
        string results =
            Utilities.CallAPI("artist.search", "", "artist=" + txtSearchArtist.Text.Trim().Replace(" ", "+"));

        // remove opensearch garbage
        if (results.Contains("opensearch"))
        {
            uint startindex = (uint)(Strings.InStr(results, "<opensearch:Query") - 1); // find starting index
            uint endindex = (uint)(Strings.InStr(results, "</opensearch:itemsPerPage>") + 25); // find ending index
            results = results.Remove((int)startindex,
                (int)(endindex - startindex)); // remove (get number of chars between start and end index)
        }

        // check for errors/populate listview
        ltvResults.Items.Clear();
        // get status
        string status = Utilities.ParseMetadata(results, "lfm status=");
        if (status.Contains("ok"))
        {
            for (byte count = 0; count <= 29; count++)
            {
                // parse for data
                string[] searchnodes = { "name", "listeners" };
                Utilities.ParseXML(results, "/lfm/results/artistmatches/artist", count, ref searchnodes);

                // add to listview
                if (searchnodes[0].Contains("ERROR: ") == false)
                {
                    ltvResults.Items.Add(searchnodes[0]); // name
                    ltvResults.Items[count].SubItems
                        .Add(Conversions.ToUInteger(searchnodes[1]).ToString("N0")); // listeners
                }
            }
        }
        else // if errors
        {
            MessageBox.Show("API ERROR: Cannot retrieve search results.", "Advanced Search", MessageBoxButtons.OK,
                MessageBoxIcon.Error);
            return;
        }

        // display message if no results found
        if (ltvResults.Items.Count == 0)
        {
            MessageBox.Show("No results found, please alter your search and try again.",
                "Search: " + txtSearchArtist.Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
            txtSearchArtist.Select();
        }
    }

    // put selected search item into final boxes
    private void SelectSearchItem(object sender, ListViewItemSelectionChangedEventArgs e)
    {
        if (ltvResults.SelectedItems.Count > 0)
        {
            txtArtist.Text = ltvResults.SelectedItems[0].Text;
            txtListeners.Text = ltvResults.SelectedItems[0].SubItems[1].Text;
        }
    }

    private void MBID(object sender, EventArgs e)
    {
        // call api
        string info = Utilities.CallAPI("artist.getInfo", "", "mbid=" + txtMBID.Text.Trim().ToLower());

        // get info or error
        string[] infonodes = { "name", "stats/listeners" };
        Utilities.ParseXML(info, "/lfm/artist", 0U, ref infonodes);
        // check for error
        if (infonodes[0].Contains("ERROR: ") == false && infonodes[0].Contains("ERROR: ") == false)
        {
            txtArtist.Text = infonodes[0];
            try
            {
                txtListeners.Text = Conversions.ToInteger(infonodes[1]).ToString("N0");
            }
            catch (Exception ex)
            {
                txtListeners.Text = "Number Error";
            }
        }
        else
        {
            MessageBox.Show("No artist was able to be found from this MBID, please try again.",
                "MBID: " + txtMBID.Text.Trim().ToLower(), MessageBoxButtons.OK, MessageBoxIcon.Error);
            txtMBID.Select();
        }
    }

    // put final box data into main form
    private void OK(object sender, EventArgs e)
    {
        if (!string.IsNullOrEmpty(txtArtist.Text) && txtArtist.Text != "N/A") // if there is data in artist box
        {
            Utilities.GoToArtist(txtArtist.Text.Trim());
            Close(); // close window
        }
        else // if there is no data entered
        {
            Close();
        }
    }

    private void Cancel(object sender, EventArgs e)
    {
        Close();
    }

    private void UserType(object sender, KeyEventArgs e)
    {
        if (ltvResults.SelectedItems.Count == 0) return;

        // clear all info
        ltvResults.SelectedItems.Clear();
        txtListeners.Text = "N/A";
        ((TextBox)sender).Clear();
    }

    private void DoubleClickItem(object sender, EventArgs e)
    {
        btnOK.PerformClick();
    }
}